README.org
#+TITLE: Omni-Log

[[https://travis-ci.org/AdrieanKhisbe/omni-log.el][file:https://travis-ci.org/AdrieanKhisbe/omni-log.el.svg]]
[[https://coveralls.io/r/AdrieanKhisbe/omni-log.el][file:https://coveralls.io/repos/AdrieanKhisbe/omni-log.el/badge.svg]]
[[http://melpa.org/#/omni-log][file:http://melpa.org/packages/omni-log-badge.svg]]
[[http://stable.melpa.org/#/omni-log][file:http://stable.melpa.org/packages/omni-log-badge.svg]]
[[https://github.com/AdrieanKhisbe/omni-log.el/tags][file:https://img.shields.io/github/tag/AdrieanKhisbe/omni-log.el.svg]]
[[http://www.gnu.org/licenses/gpl-3.0.html][http://img.shields.io/:license-gpl3-blue.svg]]

/Log All The Things!/

* Description
This library provides logging faculty.
It enable you to create logger and logger function that print message to the
/echo Area/ and append them in a dedicated buffer.

#+BEGIN_SRC emacs-lisp
  (require 'omni-log)

  (omni-log-create "my-logger")
  (log-my-logger "Hello World")
  ;; Hello World print in the echo area and *my-logger* buffer
#+END_SRC

* Usage
+ Create a logger:
  #+BEGIN_SRC emacs-lisp
    (require 'omni-log)
    (omni-log-create "ansible") ; or  (omni-log-create 'ansible)
    ;; (ansible is a "category of fictional device or technology capable of instantaneous or superluminal communication")

    (omni-log-create "ansible-wcp" '((prompt . ">"))) ; with custom prompt
  #+END_SRC
+ Use the logger:
  #+BEGIN_SRC emacs-lisp
    (log-ansible "my message") ; will output the message to *ansible* buffer
    (log-ansible "Hello %s" "World")

    (-log "ansible" "Coucou :)") ; or (-log 'ansible "Coucou :)")
  #+END_SRC
+ Modify the logger afterward
  #+BEGIN_SRC emacs-lisp
    ;; set the fading option t the logger
    (omni-log-logger-set-property my-logger 'fading t)
    (omni-log-logger-set-property my-logger 'fading nil)
    ;; destroy the logger once done
    (omni-log-kill-logger my-logger)
  #+END_SRC

** Customizing Logger Options

An omni-log can take an extra set of option in an alist to customize it's behavior.
Here is the list
- prompt :: a prefix string used in the logged messages
- fading :: boolean option to have the message fade away
- fading-delay :: delay in second after which the fading begin
- fading-duration :: duration in second of the fading effect
- centered :: are the messages centered in the echo area

** Extended Usage
 If you do not wish to create the logger function with the name =log= you could create
 a logger and use it as argument to the =log= function:

 #+BEGIN_SRC emacs-lisp
   (setq my-logger (omni-log-create-logger 'my-logger '((prompt . ">x>"))))
   (-log my-logger "See, no function log-my-logger")
 #+END_SRC

* Installation
Omni-log is available on melpa, just do a =(package-install 'omni-log)= to retrieve it

* Notes

It works but this is still a prototype. I use it on my =omni-quotes= package, but API is not frozen yet.

*If you have any remark, refactor suggestion or else, just pose an issue ;)*


CHANGELOG.md
# Change Log

All notable changes to *omni-log.el* will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased][unreleased]
##### # [0.4.0] - 2020-03-04
release of package refresh 2020 operation in pull request [#3](https://github.com/AdrieanKhisbe/omni-log.el/pull/3)
### Fixed
- Support for emacs 26
- Renamed entrypoint alias `log` to `-log` to avoid logarithmic colision

##### # [0.3.6] - 2017-09-30
### Changed
- Restore support for emacs24 by dropping frame-total-col function
##### # [0.3.5] - 2017-09-30
### Added
- Can now use a symbol as the function name
### Fixed
- Documentation hello world using wrong function
### # [0.3.4] - 2017-09-30
### Fixed
- Wrong documentation about logger
## # [0.3.3] - 2017-04-20
### Fixed
- dynamic number of step with no duration
### Changed
- renamed parameter name from centering to centered
- moment when centering occured (now in quiet-message)
## [0.3.2] - 2017-04-20
### Added
- centering of the prompt
### Changed
- dynamic number of step in the fading
## [0.3.1] - 2017-04-19
### Changed
-  omni-log-message-to-logger now return formated message
### Fixed
- fading does not occur in the buffer log
- background color of the fading prompt
## [0.3.0] - 2017-04-18
### Added
- possibility to change the config after creation
- create the logger and its function in just one call
### Changed
- Logger now accept format string and rest message
- Fading now take care of the color of the prompt

## [0.2.0] - 2017-04-18
### Added
- fading functionality to log
### Changed
- logger creator now take an alist rather than a filename

## [0.1.x] - 2015-06-04
- "Initial" stable Release

[unreleased]: https://github.com/AdrieanKhisbe/omni-log.el/compare/v0.4.0...HEAD
[0.4.0]: https://github.com/AdrieanKhisbe/omni-log.el/compare/v0.3.6...v0.4.0
[0.3.6]: https://github.com/AdrieanKhisbe/omni-log.el/compare/v0.3.5...v0.3.6
[0.3.5]: https://github.com/AdrieanKhisbe/omni-log.el/compare/v0.3.4...v0.3.5
[0.3.4]: https://github.com/AdrieanKhisbe/omni-log.el/compare/v0.3.3...v0.3.4
[0.3.3]: https://github.com/AdrieanKhisbe/omni-log.el/compare/v0.3.2...v0.3.3
[0.3.2]: https://github.com/AdrieanKhisbe/omni-log.el/compare/v0.3.1...v0.3.2
[0.3.1]: https://github.com/AdrieanKhisbe/omni-log.el/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/AdrieanKhisbe/omni-log.el/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/AdrieanKhisbe/omni-log.el/compare/v0.1.2....v0.2.0
[0.1.x]: https://github.com/AdrieanKhisbe/omni-log.el/compare/907eb8f....v0.1.2


test/test-helper.el
;; test-helper - for omni-log

;;; Code:

(require 'f)
(require 's)

(defvar omni-log-path
  (f-parent (f-this-file)))

(defvar omni-log-test-path
  (f-dirname (f-this-file)))

(defvar omni-log-root-path
  (f-parent omni-log-test-path))

(defvar omni-log-sandbox-path
  (f-expand "sandbox" omni-log-test-path))

(when (f-exists? omni-log-sandbox-path)
  (error "Something is already in %s. Check and destroy it yourself" omni-log-sandbox-path))

(defmacro within-sandbox (&rest body)
  "Evaluate BODY in an empty sandbox directory."
  `(let ((default-directory omni-log-sandbox-path))
     (when (f-exists? omni-log-sandbox-path)
       (f-delete default-directory :force))
     (f-mkdir omni-log-sandbox-path)
     ,@body
     (f-delete default-directory :force)))

(defmacro with-logger (name &rest body)
  "Evaluate BODY with a log named NAME at disposition."
  `(let ((test-logger (omni-log-create ,(eval name))))
     ,@body
     (omni-log-kill-logger test-logger)))

(defmacro with-logger-opt (name options &rest body)
  "Evaluate BODY with a log named NAME at disposition."
  `(let ((test-logger (omni-log-create ,(eval name) ,options)))
     ,@body
     (omni-log-kill-logger test-logger)))

(defun should-log (logger message)
  (with-current-buffer (omni-log-logger-buffer logger)
    (should (equal (buffer-string) message))))

(unless (s-matches? "^emacs-24\\.[12]-travis$" (or (getenv "EVM_EMACS") "local"))
  (require 'undercover)
  (undercover "*.el" "omni-log/*.el"
            (:exclude "*-test.el")
            (:send-report nil)
            (:report-file "/tmp/undercover-report.json")))
(require 'ert)
(require 'ert-async)
(require 'omni-log-logger)
(require 'omni-log)

;;; test-helper.el ends here


test/omni-log-test.el

;; ¤> Logger Creation
(ert-deftest omni-log-logger/from-logger ()
  (let ((logger '(logger "dummy-logger")))
    (should (eq logger (omni-log-logger logger)))))


(ert-deftest can-create-logger ()
  (let ((test-logger (omni-log-create-logger "ansible")))
    (should (omni-log-logger-p test-logger))
    (should (equal "ansible" (omni-log-logger-name test-logger)))
    (omni-log-kill-logger test-logger)))

(ert-deftest can-create-logger-with-symbol ()
  (let ((test-logger (omni-log-create-logger 'ansible)))
    (should (omni-log-logger-p test-logger))
    (should (equal 'ansible (omni-log-logger-symbol test-logger)))
    (should (equal "ansible" (omni-log-logger-name test-logger)))
    (omni-log-kill-logger test-logger)))

(ert-deftest can-make-logger-with-symbol ()
  (let ((test-logger (omni-log--make-logger 'ansible)))
    (should (omni-log-logger-p test-logger))
    (should (equal 'ansible (omni-log-logger-symbol test-logger)))
    (should (equal "ansible" (omni-log-logger-name test-logger)))
    (omni-log-kill-logger test-logger)))

(ert-deftest cannot-recreate-logger ()
  (let ((test-logger (omni-log-create-logger "ansible")))
    (should (equal (omni-log-create-logger "ansible") "A logger named ansible already exists"))
    (omni-log-kill-logger test-logger)))

(ert-deftest can-retrieve-logger-by-name ()
  (let ((test-logger (omni-log-create-logger "ansible")))
    (should (equal (omni-log-logger "ansible") test-logger))
    (omni-log-kill-logger test-logger)))

(ert-deftest can-retrieve-logger-by-symbol ()
  (let ((test-logger (omni-log-create-logger "ansible")))
    (should (equal (omni-log-logger 'ansible) test-logger))
    (omni-log-kill-logger test-logger)))

(ert-deftest can-create-logger-with-properties ()
  (let ((test-logger (omni-log-create-logger "ansible" '((a . "1") (b . 2)))))
    (should (omni-log-logger-p test-logger))
    (should (equal "ansible" (omni-log-logger-name test-logger)))
    (should (equal '((a . "1") (b . 2)) (omni-log-logger-properties test-logger)))
    (omni-log-kill-logger test-logger)))


(ert-deftest can-access-logger-with-properties ()
  (let ((test-logger (omni-log-create-logger "ansible" '((a . "1") (b . 2)))))
    (should (equal '((a . "1") (b . 2)) (omni-log-logger-properties test-logger)))
    (should (equal (omni-log-logger-property test-logger 'a) "1"))
    (omni-log-kill-logger test-logger)))

(ert-deftest can-access-logger-with-properties-with-default ()
  (let ((test-logger (omni-log-create-logger "ansible" '((a . "1") (b . 2)))))
    (should (equal '((a . "1") (b . 2)) (omni-log-logger-properties test-logger)))
    (should (equal (omni-log-logger-property test-logger 'a 2) "1"))
    (should (equal (omni-log-logger-property test-logger 'c 2) 2))
    (omni-log-kill-logger test-logger)))

(ert-deftest can-create-logger-and-function-in-one-call ()
  (let ((test-logger (omni-log-create "ansible" '((a . "1") (b . 2)))))
    (should (equal '((a . "1") (b . 2)) (omni-log-logger-properties test-logger)))
    (should (fboundp 'log-ansible))
    (omni-log-kill-logger test-logger)))

(ert-deftest modify-and-set-logger-property ()
  (let ((test-logger (omni-log-create-logger "ansible" '((a . "1") (b . 2)))))
    (omni-log-logger-set-property test-logger 'a 2)
    (omni-log-logger-set-property test-logger 'c 3)
    (should (equal '((a . 2) (b . 2) (c . 3)) (omni-log-logger-properties test-logger)))
    (omni-log-kill-logger test-logger)))

(ert-deftest check-logger-ko ()
  (should-error (omni-log-check-logger "toto"))
  (should-error (omni-log-check-logger '(fake-logger))))

;; ¤> Kill buffer

;; §todo: kill buffer
;; §todo: kill buffer with archive

(ert-deftest kill-no-logger ()
  (should-error
   (omni-log-kill-logger "toto")
   :type 'wrong-type-argument))

;; ¤> Logging
(ert-deftest log-with-name ()
  (with-logger "ansible"
               (-log "ansible" "Yo")
               (should-log test-logger "Yo\n")))

(ert-deftest log-with-name ()
  (with-logger "ansible"
               (-log 'ansible "Yo")
               (should-log test-logger "Yo\n")))

(ert-deftest log-with-logger ()
  (with-logger "ansible"
   (-log test-logger "Yolo!")
   (should-log test-logger "Yolo!\n")))

(ert-deftest log-no-log ()
  (should (equal (-log "unknown" "Lost Bottle")
                 "Warning (emacs): There is no logger of name unknown.")))

(ert-deftest can-log-to-created-logger ()
  (let ((test-logger (omni-log-create-logger "ansible")))
    (should (equal (omni-log-message-to-logger test-logger "42") "42"))
    (should-log test-logger "42\n")
    (omni-log-kill-logger test-logger)))

(ert-deftest can-create-logger-method ()
  (with-logger "ansible"
    (log-ansible "Working!")
    (should-log test-logger "Working!\n")))

(ert-deftest cannot-recreate-logger-method ()
  (with-logger "ansible"
    (should (equal (omni-log-create-log-function test-logger)
                   "Warning (emacs): log-ansible logging function has already been made!"))))

(ert-deftest can-create-logger-method-with-variable-name ()
  (let ((name "yo"))
    (with-logger name
      (log-yo "Working!")
      (should-log test-logger "Working!\n"))))

(ert-deftest can-log-with-rest-message ()
  (with-logger "ansible"
    (log-ansible "Working %s %d!" "on" 42)
    (should-log test-logger "Working on 42!\n")))

(ert-deftest can-log-with-prompt ()
  (with-logger-opt "prompt" '((prompt . ">>"))
                   (log-prompt "Toto")
                   (should-log test-logger ">> Toto\n")))

(ert-deftest can-log-with-fading ()
  (with-logger-opt "fading" '((fading . t))
    (modify-face 'omni-log-face "red" "blue") ; ¤hack
    (log-fading "Toto")
    (should-log test-logger "Toto\n")))

;; ¤> color utils and fading
(ert-deftest color-gradient-name ()
  (should (equal (omni-log-color-gradient-name "black" "white" 4)
                 '("black" "#333333" "#666666" "#999999" "#cccccc" "white"))))

(ert-deftest-async fading-function (done)
 (modify-face 'omni-log-face "red" "blue")
 (omni-log-quiet-fading-message "Fading" 0.1 0.5)
 (should (equal (face-attribute 'omni-log-fading-face :foreground) "red"))
 (run-at-time 1 nil
   (lambda (callback)
     ;; (should (equal (face-attribute 'omni-log-fading-face :foreground nil t) "green"))
     ;; not working cause current-message don't work
     (funcall callback)) done))


draft.el

(require 'color)
(require 'dash)


(defun color-gradient-name (start end step-number)
  (let ((gradiant (-map
   (lambda (rgb)
     (color-rgb-to-hex (nth 0 rgb) (nth 1 rgb) (nth 2 rgb)))
   (color-gradient (color-name-to-rgb start) (color-name-to-rgb end) step-number))))
    (-flatten (list start gradiant end))))

(color-gradient-name "#00ffff" "#0000ff" 100)
(color-gradient-name "#b2b2b2"  "#292b2e" 20)

(defface omni-log-face
  '((t (:inherit default
                 :weight bold :foreground "SteelBlue1")))
  "Face used for the first keyword of the tag")

(defface omni-log-fading-face
  '((t (:inherit omni-log-face
                 :foreground "SteelBlue1")))
  "Face used for the first keyword of the tag")


(defun message-fading (msg)
  (interactive "s")
  (modify-face 'omni-log-fading-face ; reset color
               (face-attribute 'omni-log-face :foreground nil t))
  (let ((timestamp (float-time)))
    (message "%s" (propertize msg 'face 'omni-log-fading-face 'log-p t
                            'timestamp timestamp))
    (-each-indexed (color-gradient-name (face-attribute 'omni-log-fading-face :foreground nil t)
                                        (face-attribute 'omni-log-fading-face :background nil t)
                                        20)
      (lambda (index color)
        (run-at-time index nil
                     (lambda (col timestamp)
                       (let ((cm (current-message)))
                         (if (and cm
                                  (get-text-property 0 'log-p cm)
                                  (equal timestamp (get-text-property 0 'timestamp cm)))
                             (modify-face 'omni-log-fading-face col))))
                     color timestamp)
        ))))

  ;; §trigger the fadding

(message-fading "toto")



;; ¤note: clignotant effet en lancant les deux en décalés.


omni-log-buffer.el
;;; omni-log-buffer.el --- Logging utilities

;; Copyright (C) 2014-2015  Adrien Becchis
;; Created:  2014-07-27
;; Author: Adrien Becchis <adriean.khisbe@live.fr>
;; Keywords: convenience, languages, tools

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; Intern structure of a logger.

;;; Building Notes:
;; for now: marker, name, buffer properties
;; ¤maybe: give a try to EIEIO: http://www.gnu.org/software/emacs/manual/html_mono/eieio.html
;; ¤see: http://nic.ferrier.me.uk/blog/2012_07/tips-and-tricks-for-emacslisp

;;; Code:

;; ¤maybe also rename intern to logger.
;;; ¤> Constructor
(defun l--make-log (name &optional properties)
  "Build a loging buffer object with NAME and eventual PROPERTIES."
  ;; ¤maybe: precise this is intern function?
  ;; ¤maybe: build properties from keyword at this level?
  (let* ((buffer (get-buffer-create  (concat "*" name "*")))
        (log (list 'log name buffer properties)))
    (with-current-buffer buffer
      (read-only-mode)) ; §later: log major mode
    log))

;;; ¤> reconnaisseur & type checkying method

(defun l-log-p (buffer) ; ¤note: maybe rename logp?
  "Return t if BUFFER is an omni-log buffer."
  ;; could be forged...[§maybe deeper check?]
  ;; §bonux: check buffer still alive? if notn maybe replace it?
  (and (consp buffer)
       (equal 'log (car buffer))))

(defun l-check-log (log)
  "Ensure That LOG is really one.  Throw exception otherwise.  Return the log."
  (if (l-log-p log)
      log
    (signal 'wrong-type-argument '(l-log-p log)))) ; ¤check message: maybe send `type-of'
;; "Provided buffer is not a log")

;; ¤note:maybe two way to call to log. registered name, or particular logbuffer. (this checkying function is for that,)

;;; ¤> accessor functions
(defun l-log-name (log)
  "Get name of LOG."
  (nth 1 (l-check-log log)))
;; ¤todo: cadr said to be coming from the `cl' package. (flycheck warning)
;; maybe use `nth' instead?

(defun l-log-buffer (log)
  "Get buffer of LOG."
  (nth 2 (l-check-log log)))

(defun l-log-properties (log)
  "Get properties of LOG."
  (nth 3 (l-check-log log)))

;; §maybe: renaming function
;; ¤tmp: test (l-log-properties a)

;; §todo: wrapper avec une qui encaspule le nom
;; §todo: get-log-or-create

;; ¤> saving fonctionnality
;; §later: (defcustom l-default-saving-interval 5
;; §see: how to handle this. (determine when to save)
;; §this shouuld go in the major mode to create! [create file.]

(provide 'omni-log-buffer)
;;; omni-log-buffer.el ends here


omni-log-logger.el
;;; omni-log-logger.el --- Logging utilities

;; Copyright (C) 2014-2017  Adrien Becchis
;; Created: 2014-07-27
;; Author: Adrien Becchis <adriean.khisbe@live.fr>
;; Keywords: convenience, languages, tools

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Building Notes:
;; for now: marker, name, buffer properties
;; ¤maybe: give a try to EIEIO: http://www.gnu.org/software/emacs/manual/html_mono/eieio.html
;; ¤see: http://nic.ferrier.me.uk/blog/2012_07/tips-and-tricks-for-emacslisp

;;; Commentary:

;; Intern structure of a logger.

;;; Code:

;;; ¤> Constructor
(defun omni-log--make-logger (name-or-symbol &optional properties)
  "Build a loging buffer object with NAME-OR-SYMBOL and eventual PROPERTIES."
  ;; ¤maybe: precise this is intern function?
  ;; ¤maybe: build properties from keyword at this level?
  (let* ((name (if (stringp name-or-symbol) name-or-symbol (symbol-name name-or-symbol)))
         (symbol (if (symbolp name-or-symbol) name-or-symbol (intern name-or-symbol)))
         (buffer (get-buffer-create  (concat "*" name "*")))
        (logger (list 'logger name symbol buffer properties)))
    (with-current-buffer buffer
      (read-only-mode)) ; §later: log major mode
    logger))

;;; ¤> reconnaisseur & type checkying method
(defun omni-log-logger-p (object) ; ¤note: maybe rename logp?
  "Return t if OBJECT is an omni-log buffer."
  ;; could be forged...[§maybe deeper check?]
  ;; §bonux: check buffer still alive? if notn maybe replace it?
  (and (consp object)
       (equal 'logger (car object))))

(defun omni-log-check-logger (logger)
  "Ensure That LOGGER is really one.  Throw exception otherwise.  Return the logger."
  (if (omni-log-logger-p logger)
      logger
    (signal 'wrong-type-argument '(omni-log-logger-p logger)))) ; ¤check message: maybe send `type-of'
;; "Provided buffer is not a log")

;; ¤note: maybe two way to call to log. registered name, or particular logbuffer. (this checkying function is for that,)

;;; ¤> accessor functions
(defun omni-log-logger-name (logger)
  "Get name of LOGGER."
  (nth 1 (omni-log-check-logger logger)))

(defun omni-log-logger-symbol (logger)
  "Get name of LOGGER."
  (nth 2 (omni-log-check-logger logger)))

(defun omni-log-logger-buffer (logger)
  "Get buffer of LOGGER."
  (nth 3 (omni-log-check-logger logger)))

(defun omni-log-logger-properties (logger)
  "Get properties of LOGGER."
  (nth 4 (omni-log-check-logger logger)))

(defun omni-log-logger-property (logger key &optional default)
  "Get KEY property of LOGGER with eventual DEFAULT."
  (let ((value (cdr (assoc key (omni-log-logger-properties logger)))))
    (if default (or value default) value)))

(defun omni-log-logger-set-property (logger key value)
  "Set KEY property of LOGGER to VALUE"
  (let ((assoc (assoc key (omni-log-logger-properties logger)))
        (props (omni-log-logger-properties logger)))
    (if assoc (setcdr assoc value)
      (progn
        (add-to-list 'props `(,key . ,value) t)
        (setf (nth 4 logger) props)))))

;; §maybe: renaming function
;; ¤tmp: test (omni-log-logger-properties a)

;; §todo: wrapper avec une qui encaspule le nom
;; §todo: get-logger-or-create

;; ¤> saving fonctionnality
;; §later: (defcustom omni-log-default-saving-interval 5
;; §see: how to handle this. (determine when to save)
;; §this should go in the major mode to create! [create file.]
;; §??? §what: did i meant by saving functionality
(provide 'omni-log-logger)
;;; omni-log-logger.el ends here


omni-log.el
;;; omni-log.el --- Logging utilities  -*- lexical-binding: t -*-

;; Copyright (C) 2014-2020  Adrien Becchis

;; Author: Adrien Becchis <adriean.khisbe@live.fr>
;; Created:  2014-07-27
;; Last-Release:  2020-03-04
;; Version: 0.4.0
;; Package-Requires: ((emacs "24") (ht "2.0") (s "1.6.1") (dash "2.13.0"))
;; Url: https://github.com/AdrieanKhisbe/omni-log.el
;; Keywords: convenience, languages, tools

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Building Notes:

;; far too early [and pretentious] to call it `l' the 'the long lost logging api' ^^

;;; Commentary:

;; Logging Utilities for packages.
;; Offer function to log messages to dedicated buffers

;;; Code:

(require 'dash)
(require 's)
(require 'ht)
(require 'color)
(require 'omni-log-logger)

(defvar omni-log-logger-index (ht)
  "Logger hash containing associating between name and logger.")

(defface omni-log-face
  '((t (:inherit default)))
  "Face for the omni-log message")

(defface omni-log-fading-face
  '((t (:inherit omni-log-face)))
  "Face for the omni-log message when fading")

(defface omni-log-prompt-face
  '((t (:inherit font-lock-keyword-face :weight bold)))
  "Face for the omni-log prompt")

(defface omni-log-fading-prompt-face
  '((t (:inherit omni-log-prompt-face)))
  "Face for the omni-log prompt when fading")

(defun omni-log-quiet-message (message &optional centered)
  "Print a MESSAGE in the loggin area without recording it in the *Messages* buffer.
Message will be CENTERED if option is provided."
  ;; inspired from eldoc
  (let ((message-log-max nil))
    ;; ¤note: centering, or right alignment should happen here.
    ;;        but choice to do so is responsability of the buffer.
    ;;        ¤maybe create another no log padding,whatever, and message-to-log would dispacth
    ;;        [object oriented programming where are you when we need you?]
    ;;        this prevent to have a rest version
    (message  (if centered (omni-log--centering message) message))))

(defun omni-log--centering (text)
  (s-center (- (frame-width) 3) text))

;; §then color. (highligh/bold: ou plus `emphasize')
;; insert color in message: log-message-with-color

(defun omni-log-logger (logger-or-name)
  "Return logger from LOGGER-OR-NAME or nil if non existing."
  (if (omni-log-logger-p logger-or-name)
      logger-or-name
    (omni-log-get-logger logger-or-name)))

(defun omni-log-get-logger (name-or-symbol)
  "Send back the eventual buffer with specified NAME."
  (ht-get omni-log-logger-index
          (if (symbolp name-or-symbol)
              (symbol-name name-or-symbol)
            name-or-symbol)
          nil))

(defun omni-log-create (name &optional properties)
  "Create and return a logger with given NAME and PROPERTIES.
take care to create the logger function"
  (let ((logger (omni-log-create-logger name properties)))
    (omni-log-create-log-function logger)
    logger))

(defun omni-log-create-logger (name &optional properties)
  "Create and return a logger with given NAME.

The logger is both registered and returned to be eventually
asigned to a variable.  An optional PROPERTIES is accepted.
§TODO: param list options

Warning will be issued if a logger with same NAME already exists."
  ;; ¤idea: keyword to signal intensity-> omni-log-logger-/name/ :info "blable"

  ;; §otherParam: filename, saving frequencing etc.
  ;; §keywordp?
  ;; §maybe: create holding var and functions? [maybe at a higher level?]
  (interactive "sName of the logger: ")
  ;; §todo: sanitize name?
  ;; §todo: then check no name conflict
  (if (omni-log-get-logger name)
      (message "A logger named %s already exists" name)
    (let* ((name (if (symbolp name) (symbol-name name) name))
           (logger (omni-log--make-logger name properties)))
      ;; §todo: check provided properties
        (ht-set! omni-log-logger-index name logger)
        logger)))

(defun omni-log-kill-logger (logger-or-name &optional archive)
  "Kill LOGGER-OR-NAME.  If ARCHIVE ask, the buffer will be renamed (and returned)."
  ;; §todo: kill attached logging methods
  (let ((logger (omni-log-logger logger-or-name)))
    (unless logger (signal 'wrong-type-argument '(omni-log-logger-p logger)))
    (let ((name (omni-log-logger-name logger)))
      (ht-remove! omni-log-logger-index name)
      ;; remove logging function (whether it has bee create of not)
      (fmakunbound (intern (concat "log-" name)))
      (if archive
          (with-current-buffer (omni-log-logger-buffer logger)
            (rename-buffer (format "%s-old" name) t))
        (kill-buffer (omni-log-logger-buffer logger))))))


(defun omni-log-create-log-function (logger)
  ;; §todo: add possibility to bypass logger name?
  "Create a function to directly append to LOGGER the given message.
This function would be named `log-' followed by logger name"
  ;;§todo: check not set!
  (let ((name (intern (concat "log-" (omni-log-logger-name logger)))))
    (if (fboundp name)
        (warn "%s logging function has already been made!" name)
      (omni-log--make-log-function name logger)))
    ;; §todo: save it in the log obxject
  )

;; §check lexical biniding
(defun omni-log--make-log-function (function-name logger)
  "Create the logging FUNCTION-NAME attached to the given LOGGER.

This is not intended for users."
  ;; ¤note: beware macro name conflict: var name must be different from the one used in log.
  (defalias function-name
    (function (lambda (message &rest args)
                (format "Log given MESSAGE to the %s logger" (omni-log-logger-name logger))
                (interactive "s")
                (apply 'omni-log-message-to-logger logger message args)))))

(defun -log (logger-or-name format-string &rest args); rest-args to do
  ;; originaly named log, but cause naming collision with native logarithmic log from calc
  "Log to specified LOGGER-OR-NAME given MESSAGE.
LOGGER-OR-NAME is either a logger or the name of the existing logger"
    (let ((logger (omni-log-logger logger-or-name)))
      (if logger
          (apply 'omni-log-message-to-logger logger format-string args)
        (warn "There is no logger of name %s." logger-or-name))))

(defun omni-log-message-to-logger (logger format-string &rest args)
  "Add to LOGGER given FORMAT-STRING and ARGS and display it in the Echo area.
Returns formatted message."
  ;; §later: evaluate message content now. and enable multi format (format style)
  (let* ((prompt-prop (omni-log-logger-property logger 'prompt))
         (prompt (if prompt-prop (concat prompt-prop " ") ""))
         (centered (omni-log-logger-property logger 'centered))
         (fading (omni-log-logger-property logger 'fading))
         (fading-delay (omni-log-logger-property logger 'fading-delay))
         (fading-duration (omni-log-logger-property logger 'fading-duration))
         (message (apply 'format format-string args))
         (message-static (format "%s%s" (propertize prompt 'face 'omni-log-prompt-face)
                                 (propertize message 'face 'omni-log-face)))
         (message-fading (format "%s%s" (propertize prompt 'face 'omni-log-fading-prompt-face)
                                 (propertize message 'face 'omni-log-fading-face))))
    (omni-log--append-to-logger (omni-log-check-logger logger) message-static)
    (if fading
        (omni-log-quiet-fading-message message-fading centered fading-delay fading-duration)
      (omni-log-quiet-message message-static centered))
    message-static))

(defun omni-log-quiet-fading-message (message &optional centered delay duration)
  "Log given MESSAGE in a fading way"
  (let* ((timestamp (float-time))
        (delay (or delay 2))
        (duration (or duration 5))
        (nstep (floor (* duration 10))))
          (modify-face 'omni-log-fading-face ; reset color
                       (face-attribute 'omni-log-face :foreground nil t))
          (modify-face 'omni-log-fading-prompt-face ; reset color
                       (face-attribute 'omni-log-prompt-face :foreground nil t))
          (omni-log-quiet-message (propertize message 'log-p t 'timestamp timestamp) centered)
          (-each-indexed
              (-zip
               (omni-log-color-gradient-name ; text color
                (let ((foreground (face-attribute 'omni-log-fading-face :foreground nil t)))
                  (if (equal foreground "unspecified-fg") "white" foreground))
                (let ((background (face-attribute 'omni-log-fading-face :background nil t)))
                  (if (or (equal background "unspecified-bg") (equal background 'unspecified)) "black" background))
                nstep)
               (omni-log-color-gradient-name ; prompt color
                (let ((foreground (face-attribute 'omni-log-fading-prompt-face :foreground nil t)))
                  (if (or (equal foreground "unspecified-fg") (equal foreground 'unspecified)) "white" foreground))
                (let ((background (face-attribute 'omni-log-fading-face :background nil t)))
                  (if (or (equal background "unspecified-bg") (equal background 'unspecified)) "black" background))
               nstep)
               )
            (lambda (index colors)
              (run-at-time (+ delay (* index (/ (float duration) nstep))) nil
                           (lambda (cols timestamp)
                             (let ((cm (current-message)))
                               (when (and cm
                                        (get-text-property 0 'log-p cm)
                                        (equal timestamp (get-text-property 0 'timestamp cm)))
                                 (modify-face 'omni-log-fading-face (car cols))
                                 (modify-face 'omni-log-fading-prompt-face (cdr cols)))))
                           colors timestamp)))))

(defun omni-log--append-to-logger (logger message)
  "Append to LOGGER given MESSAGE."
  ;; ¤note: type checking supposed to be done at a higher level
  (with-current-buffer (omni-log-logger-buffer logger)
    ;; §maybe: create a with-current-logger
    (goto-char (point-max)) ;; ¤note: maybe use some mark if the bottom of the buffer us some text or so
    (let ((inhibit-read-only t))
      (insert message)
      ;; §todo: call to special formater on message: add timestamp (maybe calling function? (if can be retrieved from namespace))
      ;; ¤note: message is supposed to be already formated. (-> color empahsize inside should be already done)
    (newline))))

(let ((omni-log--rgb-to-hex
       (if (>= emacs-major-version 26)
           (lambda (rgb) (color-rgb-to-hex (nth 0 rgb) (nth 1 rgb) (nth 2 rgb) 2))
           (lambda (rgb) (color-rgb-to-hex (nth 0 rgb) (nth 1 rgb) (nth 2 rgb))))))
  (defun omni-log-color-gradient-name (start end step-number)
    (let ((gradiant
            (-map omni-log--rgb-to-hex
              (color-gradient (color-name-to-rgb start) (color-name-to-rgb end) step-number))))
      (-flatten (list start gradiant end)))))

(provide 'omni-log)
;;; omni-log.el ends here
